### PR TITLE
Refactor of BNG tests to make them compatible with new fabric-bng pipeline

### PR DESCRIPTION
**Pull request by daniele-moro**
- compatible with pop and route and route and push functionality implemented in fabric.p4.
- compatible with new fabric-bng pipeline

### CODE REVIEW

File: tests/ptf/fabric_test.py [hunk 1]
```diff
@@ -150,6 +151,14 @@ def pkt_add_vlan(pkt, vlan_vid=10, vlan_pcp=0, dl_vlan_cfi=0):
            pkt[Ether].payload
 
 
+def pkt_add_inner_vlan(pkt, vlan_vid=10, vlan_pcp=0, dl_vlan_cfi=0):
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: Out of curiosity: what's wrong with the previous way of adding the inner vlan, i.e. calling twice  pkt_add_vlan?

File: tests/ptf/fabric_test.py [hunk 2]
```diff
@@ -540,6 +578,10 @@ def runXConnectTest(self, pkt):
         vlan_id_outer = 100
         vlan_id_inner = 200
         next_id = 99
+        # self.set_ingress_port_vlan(ingress_port=self.port1, vlan_id=vlan_id_outer,
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: Why are these lines commented out?

File: tests/ptf/fabric_test.py [hunk 3]
```diff
@@ -698,6 +740,159 @@ def runIPv4UnicastTest(self, pkt, next_hop_mac,
         testutils.verify_no_other_packets(self)
 
 
+class PopAndRouteTest(FabricTest):
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: let's this DoubleVlanTerminationTest

File: tests/ptf/fabric_test.py [hunk 4]
```diff
@@ -1248,22 +1435,21 @@ def runDownstreamV4Test(self, pkt, in_tagged, line_enabled):
...
 
         # Build expected packet from the input one, we expect it to be routed
         # and encapsulated in double VLAN tags and PPPoE.
-        exp_pkt_base = pkt_add_pppoe(pkt, type=1, code=PPPOE_CODE_SESSION_STAGE,
-                                     session_id=pppoe_session_id)
-        exp_pkt_base = pkt_add_vlan(exp_pkt_base, vlan_vid=vlan_id_inner)
-        exp_pkt_base = pkt_add_vlan(exp_pkt_base, vlan_vid=vlan_id_outer)
+        exp_pkt = pkt_add_pppoe(pkt, type=1, code=PPPOE_CODE_SESSION_STAGE, session_id=pppoe_session_id)
+        exp_pkt = pkt_add_vlan(exp_pkt, vlan_vid=vlan_id_inner)
+        exp_pkt = pkt_add_vlan(exp_pkt, vlan_vid=vlan_id_outer)
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: it's unclear why you need the new function `pkt_add_innver_vlan` if you don't use it

File: tests/ptf/fabric_test.py (hunk 1 above)
Comment (daniele-moro) [reply tests/ptf/fabric_test.py]: To add 2 VLAN to a packet you can call 2 times the `pkt_add_vlan`. The problem is when you have a packet with a VLAN already set and you want to add an inner VLAN tag, in this case you have to use the `pkt_add_inner_vlan`.

File: tests/ptf/fabric_test.py (hunk 2 above)
Comment (daniele-moro) [reply tests/ptf/fabric_test.py]: Not needed, useless code.

File: tests/ptf/fabric_test.py [hunk 5]
```diff
@@ -698,6 +740,159 @@ def runIPv4UnicastTest(self, pkt, next_hop_mac,
...
+        if Dot1Q not in pkt:
+            pkt_to_send = pkt_add_vlan(pkt_to_send, vlan_vid=vlan_id)
+            pkt_to_send = pkt_add_vlan(pkt_to_send, vlan_vid=inner_vlan_id)
+        else:
+            try:
+                pkt_to_send[Dot1Q:2]
+            except IndexError:
+                # Add the not added vlan header
+                if pkt_to_send[Dot1Q:1].vlan == vlan_id:
+                    pkt_to_send = pkt_add_inner_vlan(pkt_to_send, vlan_vid=inner_vlan_id)
```
Comment (daniele-moro) [inline tests/ptf/fabric_test.py]: @ccascone The only point in which `pkt_add_inner_vlan` is needed is here.

File: tests/ptf/fabric_test.py [hunk 6]
```diff
@@ -698,6 +751,172 @@ def runIPv4UnicastTest(self, pkt, next_hop_mac,
...
+        self.add_next_double_vlan(next_id, next_vlan_id, next_inner_vlan_id)
+
+        if exp_pkt is None:
+            # Build exp pkt using the input one.
+            exp_pkt = pkt.copy()
+            exp_pkt = pkt_add_vlan(exp_pkt, next_inner_vlan_id)
+            exp_pkt = pkt_add_vlan(exp_pkt, next_vlan_id)
+            exp_pkt = pkt_route(exp_pkt, next_hop_mac)
+            if in_tagged:
+                exp_pkt = pkt_remove_vlan(exp_pkt, in_vlan)
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: I think it's better to remove the existing VLAN tag from exp_pkt, before adding the output tags...

File: tests/ptf/fabric_test.py [hunk 7]
```diff
@@ -698,6 +751,172 @@ def runIPv4UnicastTest(self, pkt, next_hop_mac,
...
+
+        if in_tagged and not pkt_is_tagged:
+            pkt = pkt_add_vlan(pkt, vlan_vid=in_vlan)
+
+        testutils.send_packet(self, self.port1, str(pkt))
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: The pattern for all test cases is to call `testutils.send_packet` after building the exp_pkt

File: tests/ptf/fabric_test.py [hunk 8]
```diff
@@ -698,6 +751,172 @@ def runIPv4UnicastTest(self, pkt, next_hop_mac,
...
+        if exp_pkt is None:
+            # Build exp pkt using the input one.
+            exp_pkt = pkt.copy()
+            exp_pkt = pkt_route(exp_pkt, next_hop_mac)
+            exp_pkt = pkt_remove_vlan(exp_pkt)
+            exp_pkt = pkt_remove_vlan(exp_pkt)
+            if out_tagged:
+                exp_pkt = pkt_add_vlan(exp_pkt, vlan_vid=next_vlan)
+
+        testutils.send_packet(self, self.port1, str(pkt_to_send))
```
Comment (ccascone) [inline tests/ptf/fabric_test.py]: Same here, let's move `send_packet` after `exp_pkt = pkt_decrement_ttl(exp_pkt)`

File: tests/ptf/fabric_test.py [hunk 9]
```diff
@@ -172,12 +181,29 @@ def pkt_add_gtp(pkt, out_ipv4_src, out_ipv4_dst, teid):
...
+
+
 def pkt_decrement_ttl(pkt):
     if IP in pkt:
         pkt[IP].ttl -= 1
     return pkt
 
 
+def map_eth_type_to_conditions(eth_type):
+    if eth_type == ETH_TYPE_IPV4 or eth_type == 0x800:
```
[the remark this answers is not in the document]
Comment (daniele-moro) [reply tests/ptf/fabric_test.py]: No switch/case block in Python, if needed I can implement it with a dictionary, but I think it is an overkill.

[the remark this answers is not in the document]
Comment (ccascone) [reply tests/ptf/fabric_test.py]: Right... it will never stick in my mind. I think using a dictionary is fine.

**Pull request merged**